validation/custom_constraint.rst

```diff
@@ -59,11 +59,16 @@ The validator class is also simple, and only has one required method ``validate(
 
     use Symfony\Component\Validator\Constraint;
     use Symfony\Component\Validator\ConstraintValidator;
+    use Symfony\Component\Validator\Exception\UnexpectedTypeException;
 
     class ContainsAlphanumericValidator extends ConstraintValidator
     {
         public function validate($value, Constraint $constraint)
         {
+            if (!is_string($value) && !(is_object($value) && method_exists($value, '__toString'))) {
```

Comment:
Now that =I review this, I've got a comment :). If a constraint allows casting an object to null then it should do so for any scalar. What about changing `is_string` to `is_scalar`? I can see that it's done here for example https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/Constraints/EmailValidator.php#L77.

Also by convention; a constraint should not test empty value, for this a `NotNull` or `NotBlank` constraint should be used. So it misses:
```php
if (null === $value || '' === $value) {
    return;
}
```
at first.

Reply:
I disagree with the proposed `is_scalar()` change. Even the error message of the validator says: `The string "{{ string }}" contains an illegal character ...`  so this must only work with strings, right?

Reply:
I've added this:

```php
if ('' === $value) {
    return;
}
```

Should it contain a `trim($value)` call too? Thanks!

Reply:
Hmm, casting should be ok, it's about changing the representation to make it fit, then throw an exception if it's not possible. I think we should stay consistent, if we want to keep `is_string` that strict, then let's remove the `__toString()` check.

Trimming is not as it alters the given value. Fortunately this will soon be opt-in https://github.com/symfony/symfony/pull/26484.

Reply:
@HeahDude I made more changes according to your comments. Thanks for the reviews and for your patience 🙏